src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs

```diff
@@ -1895,68 +1965,96 @@ internal static int GetMDStreamVersion(RuntimeModule module)
+            // a member of Signature, whose lifetime is ensured below.
+            IntPtr typeHandle = _declaringType.GetUnderlyingNativeHandle();
+            Init(ObjectHandleOnStack.Create(ref _this),
+                pCorSig, cCorSig,
+                fieldHandle,
+                methodHandle,
+                typeHandle);
+            Debug.Assert(_returnTypeORfieldType != null);
+        }
+
         #region Constructors
         public Signature(
-            IRuntimeMethodInfo method,
+            IRuntimeMethodInfo methodHandle,
             RuntimeType[] arguments,
             RuntimeType returnType,
             CallingConventions callingConvention)
         {
-            m_pMethod = method.Value;
-            m_arguments = arguments;
-            m_returnTypeORfieldType = returnType;
-            m_managedCallingConventionAndArgIteratorFlags = (byte)callingConvention;
+            _arguments = arguments;
+            _returnTypeORfieldType = returnType;
+            _managedCallingConventionAndArgIteratorFlags = (int)callingConvention;
+            Debug.Assert((_managedCallingConventionAndArgIteratorFlags & 0xffffff00) == 0);
+            _pMethod = methodHandle.Value;
 
-            GetSignature(null, 0, default, method, null);
+            _declaringType = RuntimeMethodHandle.GetDeclaringType(_pMethod);
+            Init(null, 0, default, _pMethod);
+            GC.KeepAlive(methodHandle);
         }
 
         public Signature(IRuntimeMethodInfo methodHandle, RuntimeType declaringType)
         {
-            GetSignature(null, 0, default, methodHandle, declaringType);
+            _declaringType = declaringType;
+            Init(null, 0, default, methodHandle.Value);
+            GC.KeepAlive(methodHandle);
         }
 
         public Signature(IRuntimeFieldInfo fieldHandle, RuntimeType declaringType)
         {
-            GetSignature(null, 0, fieldHandle.Value, null, declaringType);
+            _declaringType = declaringType;
+            Init(null, 0, fieldHandle.Value, default);
             GC.KeepAlive(fieldHandle);
         }
 
         public Signature(void* pCorSig, int cCorSig, RuntimeType declaringType)
         {
-            GetSignature(pCorSig, cCorSig, default, null, declaringType);
+            _declaringType = declaringType;
+            Init(pCorSig, cCorSig, default, default);
         }
         #endregion
 
         #region Internal Members
-        internal CallingConventions CallingConvention => (CallingConventions)(byte)m_managedCallingConventionAndArgIteratorFlags;
-        internal RuntimeType[] Arguments => m_arguments;
-        internal RuntimeType ReturnType => m_returnTypeORfieldType;
-        internal RuntimeType FieldType => m_returnTypeORfieldType;
+        internal CallingConventions CallingConvention => (CallingConventions)(_managedCallingConventionAndArgIteratorFlags & 0xff);
+        internal RuntimeType[] Arguments => _arguments ?? [];
```

Comment:
Should we rather assert that  `arguments` is non-null here?

Reply:
> It should be a bug if Arguments property is called on a field signature.

Right. Should I throw an `InvalidOperationException` instead? I'm also fine just asserting, but I'm not sure the trade off between assert vs exception when we really don't expect the case.

Reply:
This is internal property. Assert should be fine. It is good enough to assert to catch bugs in our code.